S4.Rmd

```diff
@@ -52,6 +52,21 @@ Slots and contains can specify the names of S4 classes, S3 classes (if registere
 
 To create a class, you call `setClass()`, supplying these three properties. Lets make this concrete with an example. Here we create two classes: a person with character `name` and numeric `age`, and an `Employee` that inherits slots and methods from `Person`, adding an additional `boss` slot that must be a `Person`.
 
+<!-- FIXME: the underlying model here is of two tables 'Person' and 'Employee' in a database. The 'isa' relationship is an implicit join. You've modelled rows, which means that each instance of the class is a scalar. But R works well on vectors. Effective S4 would model columns, and each instance would contain vectors -- 'People' and 'Employees', though since R uses singular integer(), etc, then 'Person' and 'Employee' but with the understanding that these are of length 0, 1, ...
```

Comment:
Hmmm, I need to think about this more.  If I do continue down the vector route, what do I need to to make `Person` behave as a data frame column?

But I might steer away from this example altogether because a person is an object in the real world, and so reference semantics seem more natural.

Reply:
I think I'll redo the basic example here, but if S4 objects can't easily go into a data frame, why not make the objects scalars and put them into list columns? 

But maybe Person is ok for a very simple example. A more complex example (e.g. a linear model) would be more clearly scalar. I'll keep thinking about it.

Reply:
It's an efficiency issue, where one ends up iterating (over objects) instead of vectorizing (slots).

Probably a more complex model is like multiple tables in a relational data base, with the S4 object managing the relations.

Probably the idea that an S4 object is a collection of vectors is somehow encouraging 'best practice', and of course there are use cases where a list-of-scalar-objects is appropriate and performance issues are not important.

That an S4 object  can't represent a column in a data.frame is viewed (in the Bioc devel community, I guess) as a serious implementation limitation of data.frame; really, a 'column' should be defined in terms of an API that consists of a few simple methods -- `length()`, `[`, etc.; this is how `S4Vectors::DataFrame()` is implemented.

Reply:
Ok, I'll emphasise that S4 objects should be vectorised where possible.

Has there been any attempt to improve R's data frames along those lines?

Should we at least make tibbles compatible with S4 objects?

Reply:
Making tibbles compatible with S4 objects as columns would be awesome. I defer to @lawremi and @hpages for expertise.